src/bitmap/multiops.rs

```diff
@@ -0,0 +1,434 @@
+        Ok(RoaringBitmap::new())
+    }
+}
+
+#[inline]
+fn try_multi_and_ref<'a, E>(
+    bitmaps: impl IntoIterator<Item = Result<&'a RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    let mut iter = bitmaps.into_iter();
+
+    // We're going to take a bunch of elements at the start of the iterator and sort
+    // them to reduce the size of our bitmap faster.
+    let mut start = collect_starting_elements(iter.by_ref())?;
+    start.sort_unstable_by_key(|bitmap| bitmap.containers.len());
+    let mut start = start.into_iter();
+
+    if let Some(mut lhs) = start.next().cloned() {
+        for rhs in start.into_iter().map(Ok).chain(iter) {
+            if lhs.is_empty() {
+                return Ok(lhs);
+            }
+            lhs &= rhs?;
+        }
+        Ok(lhs)
+    } else {
+        Ok(RoaringBitmap::new())
+    }
+}
+
+#[inline]
+fn try_multi_sub_owned<E>(
+    bitmaps: impl IntoIterator<Item = Result<RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    let mut iter = bitmaps.into_iter();
+    match iter.next().transpose()? {
+        Some(mut lhs) => {
+            for rhs in iter {
+                if lhs.is_empty() {
+                    return Ok(lhs);
+                }
+                lhs -= rhs?;
+            }
+            Ok(lhs)
+        }
+        None => Ok(RoaringBitmap::default()),
+    }
+}
+
+#[inline]
+fn try_multi_sub_ref<'a, E>(
+    bitmaps: impl IntoIterator<Item = Result<&'a RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    let mut iter = bitmaps.into_iter();
+    match iter.next().transpose()?.cloned() {
+        Some(mut lhs) => {
+            for rhs in iter {
+                if lhs.is_empty() {
+                    return Ok(lhs);
+                }
+                lhs -= rhs?;
+            }
+
+            Ok(lhs)
+        }
+        None => Ok(RoaringBitmap::default()),
+    }
+}
+
+#[inline]
+fn try_multi_or_owned<E>(
+    bitmaps: impl IntoIterator<Item = Result<RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    let mut iter = bitmaps.into_iter();
+
+    // We're going to take a bunch of elements at the start of the iterator and
+    // move the biggest one first to grow faster.
+    let mut start = collect_starting_elements(iter.by_ref())?;
+    start.sort_unstable_by_key(|bitmap| Reverse(bitmap.containers.len()));
+    let start_size = start.len();
+    let mut start = start.into_iter();
+
+    let mut containers = if let Some(c) = start.next() {
+        if c.is_empty() {
+            // everything must be empty if the max is empty
+            start.by_ref().skip(start_size).next();
+        }
+        c.containers
+    } else {
+        return Ok(RoaringBitmap::new());
+    };
+
+    for bitmap in start.into_iter().map(Ok).chain(iter) {
+        merge_container_owned(&mut containers, bitmap?.containers, BitOrAssign::bitor_assign);
+    }
+
+    RetainMut::retain_mut(&mut containers, |container| {
+        container.ensure_correct_store();
+        container.len() > 0
```

Comment:
wouldn't it be more performant to check if we want to keep this store than converting it to the correct store type before?

Reply:
Noice we won another ±5%